src/uct/sm/mm/mm_ep.c

```diff
@@ -14,39 +14,79 @@ SGLIB_DEFINE_HASHED_CONTAINER_FUNCTIONS(uct_mm_remote_seg_t,
-        ret = sendto(iface->signal_fd, &sig, sizeof(sig), 0,
-                     (const struct sockaddr*)&ep->cached_signal_sockaddr,
-                     ep->cached_signal_addrlen);
-        if (ret >= 0) {
-            ucs_assert(ret == sizeof(sig));
-            return UCS_OK;
-        } else if (errno == EAGAIN) {
-            /* If sending a signal has failed, retry.
-             * Note the by default the receiver might have a limited backlog,
-             * on Linux systems it is net.unix.max_dgram_qlen (10 by default).
-             */
-            uct_mm_iface_recv_messages(iface);
+    ret = sendto(iface->signal_fd, &sig, sizeof(sig), 0,
+                 (const struct sockaddr*)&ep->cached_signal_sockaddr,
+                 ep->cached_signal_addrlen);
+    if (ret >= 0) {
+        ucs_assert(ret == sizeof(sig));
+        ucs_debug("Sent %sconnect from socket %d to %p",
+                  (sig == UCT_MM_IFACE_SIGNAL_CONNECT) ? "" : "dis", iface->signal_fd,
+                  (const struct sockaddr*)&ep->cached_signal_sockaddr);
+
+        if (ep->cbq_elem_on) {
+            uct_worker_slowpath_progress_unregister(iface->super.worker, &ep->cbq_elem);
+            ep->cbq_elem_on = 0;
+        }
+
+        /* point the ep->fifo_ctl to the remote fifo.
+         * it's an aligned pointer to the beginning of the ctl struct in the remote FIFO */
+        ep->fifo_ctl    = (uct_mm_fifo_ctl_t*) ucs_align_up_pow2
+                          ((uintptr_t) ep->mapped_desc.address , UCS_SYS_CACHE_LINE_SIZE);
+        ep->cached_tail = ep->fifo_ctl->tail;
+
+        return UCS_OK;
+    } else if (errno == EAGAIN) {
+        /* If sending a signal has failed, retry.
+         * Note that by default the receiver might have a limited backlog,
+         * on Linux systems it is net.unix.max_dgram_qlen (10 by default).
+         */
+        ucs_debug("Failed to send %sconnect from socket %d to %p",
+                  (sig == UCT_MM_IFACE_SIGNAL_CONNECT) ? "" : "dis", iface->signal_fd,
+                  (const struct sockaddr*)&ep->cached_signal_sockaddr);
+
+        /* If sending the Connect message failed with EAGAIN, try again later.
+         * Don't keep trying now in a loop since this may cause a deadlock which
+         * prevents the reading of incoming messages which blocks the remote sender.
+         * Add the sending attempt as a callback to a slow progress.
+         * If the attempt was sending a Disconnect message, don't try it again.
+         * All Disconnect does is remove the progress function from the progress
+         * queue. This will be done anyway (later) from the iface clean-up function. */
+        if ((!ep->cbq_elem_on) && (sig == UCT_MM_IFACE_SIGNAL_CONNECT)) {
+             ep->cbq_elem.cb = uct_mm_ep_signal_remote_slow_path_callback;
+             uct_worker_slowpath_progress_register(iface->super.worker, &ep->cbq_elem);
+             ep->cbq_elem_on = 1;
+        }
+        uct_mm_iface_recv_messages(iface);
```

Comment:
now we don't need to call this because we don't block (also make this function static)